Initial integration of Prefect for task & workflow execution

dgarros opened the pull request:
This PR integrates Prefect inside Infrahub to provide a more advanced task & workflow execution engine.
Our current engine is powerful to execute tasks but it doesn't provide the level of visibility and troubleshooting that we need and it doesn't support workflows composed of multiple tasks.

Prefect is build on the concepts of task, flow, deployment and workers where each task will be executed in a dedicated process, either locally or on a remote system (K8s) depending on the type of workers that have been deployed. this flexibility will be very handy for managing user provided code in the near future.

This PR is laying down the base framework to integrate Prefect into Infrahub by adding prefect to our current list of services. A wrapper has been added around the execution of `flows` to automatically between local execution or remote execution based on Infrahub Settings, this gives us the same flexibility as the current BusSimulator.

For now only 3 tasks have been converted :
- send_telemetry_push
- send_webhook
- transform_render_jinja2_template

My goal is to merge this PR early and continue the work in `develop`, I'll open additional issues to track the remaining work 

### Async Worker
However, the default mode of isolated environment isn't working well with our current architecture that relies on a shared registry to access the internal schema. 
In order for Prefect to support our internal task as well, this PR includes a custom Prefect Worker that execute async tasks inside the main async loop, with access to the shared resources. This behavior is very similar to what we are currently doing within the git agent and it should allow us to migrate most tasks to prefect.

dgarros committed:
Initial integration of Prefect

dgarros committed:
Move git integration fixtures into upper level

dgarros committed:
Add catalogue of workflow
Add driver to control execution

dgarros committed:
Move worker adapter to services

dgarros committed:
Update workflow catalogue format, add supprot for cron, fix worker name

dgarros committed:
Add support for local cache
Fix WorkflowWorkerExecution

dgarros committed:
Remove test_rfile_endpoint test

dgarros committed:
Add task-manager to nats

dgarros committed:
Fix yaml format

dgarros committed:
Add pytest-env

dgarros committed:
remove autouse on prefect_test

dgarros committed:
Update prefect to 3.0

dgarros committed:
Fix yamllint

dgarros committed:
Fir task format

dgarros committed:
update docker compose for test

dgarros committed:
Fix format

dgarros committed:
Add type to api_endpoint

dgarros committed:
Add prefect autouse fixture for all integration tests

dgarros committed:
Migrate prefect to postgres

dgarros committed:
Fix yaml

dgarros committed:
Add pytest-env to python sdk

dgarros committed:
Add fixture for prefect during python sdk integration tests

dgarros committed:
Fix typing

dgarros committed:
Add workflow_db storage

dgarros committed:
Add  task-manager-db for NATS

fatih-acar commented on docker-compose.yml:
```diff
@@ -145,6 +145,39 @@ services:
+      - task-manager-db
+
+  task-manager-db:
+    image: postgres:16-alpine
```
```suggestion
    image: postgres:16-alpine
    restart: unless-stopped
```

fatih-acar committed:
feat(helm): add task runner and prefect dependency

Signed-off-by: Fatih Acar <fatih@opsmill.com>

dgarros committed:
Add volume to nats docker compose

dgarros committed:
Add restart: unless-stopped to postgres, remove port

ogenstad commented on backend/infrahub/message_bus/operations/send/telemetry.py:
```diff
@@ -112,3 +118,31 @@ async def push(
+    """Send the telemetry data to the specified URL, using HTTP POST."""
+    async with httpx.AsyncClient() as client:
+        response = await client.post(url=url, json=payload)
+        response.raise_for_status()
```
For new http calls I'd suggest that we start using `service.http` and consider what to start to raise within those functions instead.

ogenstad commented on backend/infrahub/core/initialization.py:
```diff
@@ -130,6 +155,8 @@ async def initialization(db: InfrahubDatabase) -> None:
         else:
             log.warning("The database index manager hasn't been initialized.")
 
+        await initialize_tasks()
```
Can we move this so that it instead happens within `await service.initialize()`?

dgarros replied:
Great idea, I moved it

ogenstad commented on backend/infrahub/services/adapters/workflow/__init__.py:
```diff
@@ -0,0 +1,18 @@
+        workflow: WorkflowDefinition | None = None,
+        function: Callable[..., Awaitable[Return]] | None = None,
+        **kwargs: dict[str, Any],
+    ) -> Any:
```
The return code here differs from the other ones for example `WorkflowLocalExecution.execute()`

dgarros replied:
fixed

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -0,0 +1,131 @@
+            database=database,
+            message_bus=message_bus,
+            component_type=ComponentType.GIT_AGENT,
+        )
```
We aren't adding the workflow here, will that cause issues if we try to use the workflow in a consistent way using the service object within the various operations?

ogenstad commented on backend/infrahub/workflows/models.py:
```diff
@@ -0,0 +1,45 @@
+
+    @property
+    def full_name(self) -> str:
+        return f"{self.name}/{self.name}"
```
Is this correct? I.e. shoule full_name be self.name repeated twice?

dgarros replied:
Yes this is surprisingly correct ... we might be able to improve that but for now this is correct

ogenstad commented on development/Dockerfile:
```diff
@@ -72,7 +72,7 @@ RUN git config --global user.name "Infrahub" && \
     git config --global credential.helper /usr/local/bin/infrahub-git-credential
 
-RUN mkdir -p /opt/infrahub/git /opt/infrahub/storage /opt/infrahub/source /opt/infrahub/frontend/app/dist
+RUN mkdir -p /opt/infrahub/git /opt/infrahub/storage /opt/infrahub/source /opt/infrahub/frontend/app/dist /opt/infrahub/workflow
```
Does the /opt/infrahub/workflow folder need to be shared in any way or is it just for temporary storage? I.e. in the root level compose file.

dgarros replied:
it's a temporary storage, I would like to implement a driver to use Redis/NATS instead

ogenstad commented on backend/infrahub/api/transformation.py:
```diff
@@ -146,5 +144,8 @@ async def transform_jinja2(
-    return PlainTextResponse(content=response.data.rendered_template)
+    service: InfrahubServices = request.app.state.service
+
+    response: str = await service.workflow.execute(workflow=TRANSFORM_JINJA2_RENDER, message=message)  # type: ignore[arg-type]
```
I think it feels a bit verbose that we need to send in both the workflow and the message here. I think it would be nice if we could have a mapping table internally within `.execute()` that provided the workflow for a given message in the same way that `service.sent()` provide the routing_key today.

dgarros replied:
point taken, let's revisit that in a future PR

ogenstad reviewed (approved):
LGTM. I added a few comments / questions and I do think that it would be nice if we could simplify the expectations for workflow.execute() as per the comments on that part of the code.

dgarros committed:
Move tasks initialization to workflow service

fatih-acar commented on development/docker-compose-deps.yml:
```diff
@@ -19,3 +19,35 @@ services:
+    command: prefect server start --host 0.0.0.0 --ui
+    environment:
+      PREFECT_API_DATABASE_CONNECTION_URL: postgresql+asyncpg://postgres:postgres@task-manager-db:5432/prefect
+      PREFECT_LOCAL_STORAGE_PATH: ${PREFECT_LOCAL_STORAGE_PATH:-/opt/prefect/}
```
Shouldn't we also use the same docker volume as the other containers? Otherwise I think we cannot see run states inside the Prefect UI.

fatih-acar commented on backend/infrahub/services/adapters/workflow/worker.py:
```diff
@@ -0,0 +1,71 @@
+            response: FlowRun = await run_deployment(name=workflow.full_name, parameters=kwargs or {})  # type: ignore[return-value, misc]
+            if not response.state:
+                raise RuntimeError("Unable to read state from the response")
+            result = response.state.result()
```
Did you try:

```suggestion
            result = await response.state.result(fetch=True)
```

Instead of doing the below stuff yourself?

dgarros replied:
Indeed much simpler

dgarros committed:
Update worker execution code
Ensure prefect worker read the proper config file
Set transform_render_jinja2_template as persistent

fatih-acar commented:
I added the required helm chart changes: note that we now require a new shared storage for the workflow persistence feature. As mentioned in a previous comment, implementing a new storage driver based on the actual caching layer we have for that would ease deployments.
